internal/yamlalias/wrap.go

```diff
@@ -0,0 +1,93 @@
+// Package yamlalias adds support for aliases when parsing YAML.
+//
+// Aliases are specified as a comma-separated list specified via
+// "yaml-aliases" struct tag.
+//
+// Example:
+//
+//   type Config struct {
+//     UserName string `yaml:"username" yaml-aliases:"userName,user-name"`
+//   }
+//
+// This will parse the following YAML examples in the same way:
+//   {"username": "John Doe"}
+//   {"userName": "John Doe"}
+//   {"user-name": "John Doe"}
+//
+// Note: It only adds aliases for top-level fields in the struct.
+package yamlalias
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+)
+
+// Unmarshal is a wrapper for yaml.Unmarshal but adds
+// support for any yaml-aliases specified in out.
+func Unmarshal(in []byte, out interface{}) error {
+	return yaml.Unmarshal(in, addAliases(out))
+}
+
+// UnmarshalStrict is a wrapper for yaml.UnmarshalStrict but adds
+// support for any yaml-aliases specified in out.
+func UnmarshalStrict(in []byte, out interface{}) error {
+	return yaml.UnmarshalStrict(in, addAliases(out))
+}
+
+// addAliases generates a new struct that contains all exported fields
+// from the passed in struct, but adds additional fields for any
+// aliases.
+func addAliases(v interface{}) interface{} {
```

Comment:
I think it's a bit difficult to follow what this is doing in Go terms. Some comments referencing examples would help.

I *think* what this does is, given:

```
var cfg struct {
  UserName string `yaml:"username" yaml-aliases:"userName,user-name"`
}
```

It generates:

```
var dest struct {
  UserName *string `yaml:"username"`
  UserNameYamlAlias1 *string `yaml:"userName"`
  UserNameYamlAlias2 *string `yaml:"user-name"`
}
dest.UserName = &cfg.UserName
dest.UserNameYamlAlias1 = &cfg.UserName
dest.UserNameYamlAlias2 = &cfg.UserName
```

If that's correct, it's surprising to me that the YAML library respects the
pointer and fills the right destination rather than replacing the pointer.
Worth calling out?

Reply:
That's exactly right, added documentation as suggested. Here's the repro for yaml respecting existing pointers:
https://play.golang.org/p/F1AoTeUVkAW

I've added a link to a gist in the docs, but open to alternative suggestions.

The test validate all the behaviour in an end-to-end fashion, so not sure it's worth adding a test specific just for the YAML part.

Reply:
A test for just the YAML part isn't necessary. I just wasn't expecting this to work.
Looking at the yaml.Unmarshal docs, it appears deliberate so we can rely on it.

> If an internal pointer within a struct is not initialized, the yaml package will initialize it [..]

https://godoc.org/gopkg.in/yaml.v2#Unmarshal